src/System.Management.Automation/engine/parser/TypeResolver.cs

```diff
@@ -117,8 +117,10 @@ public AmbiguousTypeException(TypeName typeName, IEnumerable<string> candidates)
                             }
                         }
                     }
-                    else
+                    else if (!assembly.IsDynamic)
```

Comment:
Within the same load context, only one assembly with the same identity can be loaded. When you try to load the second, the first loaded assembly will be returned.
```
PS:3> $m = [System.Reflection.Assembly]::LoadFrom("F:\MyTypeTest\MyTypeTest.dll")
PS:4> $n = [System.Reflection.Assembly]::LoadFrom("F:\MyCmdletTest\MyTypeTest.dll")
PS:5> [System.AppDomain]::CurrentDomain.GetAssemblies() | ? FullName -like "MyTypeTest*"

GAC    Version        Location
---    -------        --------
False  v4.0.30319     F:\MyTypeTest\MyTypeTest.dll

PS:6> [System.Object]::ReferenceEquals($m, $n)
True
```
powershell core only has one load context (_unless the extension code spins up another assembly load context, in which case we cannot get the loaded assemblies from it anyways_).

**The tricky part is in windows powershell when assemblies are loaded into different load context.**
```
PS:1> $m = [System.Reflection.Assembly]::LoadFrom("F:\MyTypeTest\MyTypeTest.dll")
PS:2> $n = [System.Reflection.Assembly]::LoadFile("F:\MyCmdletTest\MyTypeTest.dll")
PS:3> [System.Object]::ReferenceEquals($m, $n)
False
PS:4> [System.AppDomain]::CurrentDomain.GetAssemblies() | ? FullName -like "MyTypeTest*"

GAC    Version        Location
---    -------        --------
False  v4.0.30319     F:\MyTypeTest\MyTypeTest.dll
False  v4.0.30319     F:\MyCmdletTest\MyTypeTest.dll
```

I think the safest way should be putting the assembly itself to the HashSet, instead of having something that represents the assembly.